### DISCUSSION TITLE
Can we have the ggwave library to support Arduino?

### DISCUSSION THREAD

**Discussion by junwoo** [ggerganov/ggwave#62] [General]
we used to use Chrip Arduino as data over sound solution, can we have ggwave library as alternative solution as Chrip was sold and changed to closed project.

I have tried ggwave, it works perfect between different platforms, it will be great to have the library support on microchip.

Thank you.

**Comment by ggerganov** [comment]
Sure - let's do it!

Can you tell me what hardware you use for your project? ~And operating system?~

**Reply by junwoo** [reply]
Thanks for the replying, the hardware is all Arduino board like Arduino UNO, Arduino Nano, but recently I found ESP 8266 is popular and cheap solution, the MCU is built in with Wifi, it can also use same Arduino IDE and code.

**Reply by ggerganov** [reply]
Do the Arduino boards have built-in mics or speakers?
I'm not very familiar yet - will try to get one and experiment with it.

**Reply by junwoo** [reply]
Bad me, I almost forget the microphone.
Yes, it supports microphone module, pls check the link https://diyi0t.com/sound-sensor-arduino-esp8266-esp32/

**Reply by tigercoxgithub** [reply]
> Do the Arduino boards have built-in mics or speakers?
> 
> I'm not very familiar yet - will try to get one and experiment with it.

Hello, did you get any further with your arduino implementation. I am very interested in this :)

**Reply by ggerganov** [reply]
Hi, I now have the hardware - Arduino UNO + speaker. I haven't started porting the lib yet.
I think the transmission part would be possible to implement, but I'm not sure about the receiving part.

In order to implement receiving + decoding I need to develop real-time audio capture through a mic and efficient FFT implementation. If anyone can provide references to similar Arduino projects for real-time audio capture and processing, it would help a lot.

**Reply by ggerganov** [reply]
Here are some preliminary experiments with a hard-coded message -- looking good so far:

https://user-images.githubusercontent.com/1991296/166509238-42ee7b2f-72bb-4d89-95c9-f424a1f656f6.mp4

**Comment by ggerganov** [comment]
The Tx port is ready:

https://user-images.githubusercontent.com/1991296/167311587-3239f133-c134-4274-9b51-9d9f2d2e8b08.mp4

Let's see if we can make the Arduino also receive some data!

**Reply by junwoo** [reply]
that is a great job.

I found some projects which related with Arduino + Mic,  I don't if we can borrow some concept from them
https://www.circuitbasics.com/how-to-use-microphones-on-the-arduino/
https://www.hackster.io/mvtdesign/wifi-microphone-with-arduino-and-esp8266-updated-69934d

**Comment by dcmckay13** [comment]
So this is one that I've been looking into. It uses an Adafruit Huzzah 32 a Adafruit ADC:MCP3008 
and a MEM microphone. I'm not exactly sure why I couldn't get mine working bit I suspect that  if I fried my MCP300, I am not really sure why the 10bit chip is required as the Feather has multiple 12-bit ADCs. <https://github.com/weckbach/AstroMech>

**Reply by dcmckay13** [reply]
The end goal for me (before I hit a wall with the programming) was to use this other esp32 from Lily/TTGO with the built in screen to display the incoming transmission <https://www.amazon.com/LILYGO-T-Display-Arduino-Development-CH9102F/dp/B099MPFJ9M/ref=sr_1_1?m=A3P8OJPGEGCZAI&qid=1652567881&s=merchant-items&sr=1-1>

**Reply by dcmckay13** [reply]
There is another that uses an Arduino and some other compennets including a hypersonic microphone. This version seems really finicy because requires a lot more components and has to be "tuned-in" via a potentiometer. From what I've read the ADC (in both devices) needs to have a sampling rate at least twice as fast as the highest detectable frequency.   https://hackaday.com/2018/06/01/dead-simple-ultrasonic-data-communication.

**Reply by dcmckay13** [reply]
If you need to add error correction I guess Reed-Solomon error correction is the standard that many Data-Over-Sound projects use

**Reply by ggerganov** [reply]
Thanks for the links - will take a look at some point in more details.
Btw, `ggwave` already uses `Reed-Solomon` ECC.

**Comment by ggerganov** [comment]
Good news everyone! I was able to port the Rx part of `ggwave` for Arduino.
I now have the `Arduino Nano RP2040 Connect` receiving short audio messages :tada:

![IMG_0262](https://user-images.githubusercontent.com/1991296/168490405-b4d1cc72-a0ce-418f-9677-7e01d2483764.jpg)

The cool thing is that this board has a built-in microphone so you don't need any extra hardware except the actual Arduino.

The maximum supported sample rate seems to be 20000 Hz, so you won't be able to use the existing `ggwave` protocols. Instead, I plan to introduce Arduino-specific protocols which will be available in the non-Arduino library as well.

I'm still trying to figure out the optimal parameters for the transmission, but so far I am satisfied with the robustness. It's not as good as the phone or PC mics, but it should be enough for most applications. As long as the transmitter is within 40-50 cm it works great.

The main issue currently is that `ggwave` currently uses too much memory. Need to reduce the dynamic memory allocations as much as possible.

The code is currently available on the `arduino` branch https://github.com/ggerganov/ggwave/pull/65, but it is a total mess. I was in a hurry to make the PoC. Next couple of weeks I won't have much time to work on this, so expect official release some time after that.

So glad I got this working. This is going to be awesome!

**Comment by ggerganov** [comment]
The Arduino port of `ggwave` is 99% complete. So far, I got it working with the following microcontrollers:

- `Arduino UNO` (Tx only)
- `Arduino RP2040` (Tx + Rx, overall very good results)
- `ESP32` (Rx, Tx not tested yet), using `MAX9814` microphone (got a few other mics to test)
- `ESP8266` (have the chip, not tested yet)

Checkout #65 for instructions of how to use it. The PR is almost a complete rewrite of `ggwave` and includes a lot of improvements. I will be merging this soon to the `master` branch, but first I want to improve the Arduino / ESP examples and probably get some feedback from other people too.

If you want to give it a try, do the following steps:
- checkout the latest `arduino` branch
- run `cmake` and `make`
- in a terminal on your computer, start the `ggwave-cli` tool like this: `./bin/ggwave-cli -t11 -l16 -s`
- using Arduino IDE, open one of the `arduino-rx` or `esp32-rx` examples (depending on what chip you want to use) and upload the sketch to your chip
- for each example, there is a photo of a sample board configuration included in the PR
- observe the Serial output of your microcontroller and see if it is able to receive messages from your computer's speakers using the `ggwave-cli` tool

I've been doing a lot of tests and using the new `MT` protocols I am able to transfer reliably short message to/from the microcontrollers.
If some of you give this a try, please let me know how it went.

**Comment by ggerganov** [comment]
The port is now completely ready! Here is a short demonstration of Arudino, ESP32 and Android phone communicating via sound:

https://user-images.githubusercontent.com/1991296/177212762-09182300-2291-4144-affd-0cc5cb6d20ba.mp4

[Watch higher quality on YouTube](https://youtu.be/aj_GLBtU3Vw)

The PR code is now merged in the `master` branch. If you want to try it, checkout the respective examples in the [examples](https://github.com/ggerganov/ggwave/tree/master/examples) folder. Any feedback will be much appreciated!

**Comment by ggerganov** [comment]
`ggwave` is now available as Arduino library:

https://github.com/ggerganov/ggwave-arduino

<img width="1018" alt="image" src="https://user-images.githubusercontent.com/1991296/177872493-2878d4d4-543a-4f74-95db-7fb2d5f074be.png">
